org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/typeHierarchy/TypeHierarchyView.java

```diff
@@ -11,32 +11,343 @@
+
+	class SymbolsContainer {
+		private final SymbolsModel symbolsModel;
+		private volatile boolean isDirty = true;
+		private boolean temporaryLoadedDocument = false;
+		private IFile file;
+
+		SymbolsContainer(IFile file) {
+			this.symbolsModel = new SymbolsModel();
+			setFile(file);
+		}
+
+		public IDocument getDocument() {
+			var document = LSPEclipseUtils.getExistingDocument(file);
+			if (document == null) {
+				document = LSPEclipseUtils.getDocument(file);
+				temporaryLoadedDocument = document != null;
+			}
+			return document;
+		}
+
+		public void setFile(IFile file) {
+			this.file = file;
+			this.symbolsModel.setUri(file.getLocationURI());
+		}
+
+		public void dispose() {
+			if (temporaryLoadedDocument) {
+				try {
+					FileBuffers.getTextFileBufferManager().disconnect(file.getFullPath(), LocationKind.IFILE,
+							new NullProgressMonitor());
+				} catch (CoreException e) {
+					LanguageServerPlugin.logError(e);
+				}
+			}
+		}
+	}
 
-public class TypeHierarchyView extends CallHierarchyView {
 	public static final String ID = "org.eclipse.lsp4e.operations.typeHierarchy.TypeHierarchyView"; //$NON-NLS-1$
 	TypeHierarchyViewContentProvider contentProvider = new TypeHierarchyViewContentProvider();
+	DecoratingStyledCellLabelProvider symbolsLabelProvider = new DecoratingStyledCellLabelProvider(new SymbolsLabelProvider(), PlatformUI.getWorkbench().getDecoratorManager().getLabelDecorator(), DecorationContext.DEFAULT_CONTEXT);
+	// widgets
+	private PageBook pagebook;
+	private SashForm splitter;
+	private ViewForm memberViewForm;
+	private CLabel memberLabel;
+	private Label infoText;
+
+	// viewers
+	private TableViewer memberViewer;
+	protected TreeViewer treeViewer;
+
+	private volatile CompletableFuture<List<Either<SymbolInformation, DocumentSymbol>>> symbols;
+	private HashMap<URI, SymbolsContainer> cachedSymbols = new HashMap<>();
+	private IDocument document;
+	private volatile String typeName;
+
+	private final IFileBufferListener fileBufferListener = new FileBufferListenerAdapter() {
+		@Override
+		public void dirtyStateChanged(IFileBuffer buffer, boolean isDirty) {
+			if (isDirty) {
+				// check if this file has been cached:
+				var cachedSymbol = getSymbolsContainer(buffer);
+				if (cachedSymbol != null) {
+					cachedSymbol.isDirty = true;
+				}
+			}
+		}
+
+		@Override
+		public void underlyingFileMoved(IFileBuffer buffer, IPath path) {
+			var symbolsContainer = getSymbolsContainer(buffer);
+			// check if this file has been cached:
+			if (symbolsContainer != null) {
+				var file = LSPEclipseUtils.getFile(path);
+				if (file != null) {
+					//update old cache:
+					symbolsContainer.setFile(file);
+					//create new cache element under new URI
+					cachedSymbols.put(file.getLocationURI(), new SymbolsContainer(file));
+				}
+			}
+		}
+
+		private SymbolsContainer getSymbolsContainer(IFileBuffer buffer) {
+			if (buffer != null) {
+				return cachedSymbols.get(LSPEclipseUtils.toUri(buffer));
```

Comment:
LSPEclipseUtils.toUri(buffer) could return null, I think we need to check that before accessing the cachedSymbols map.

Reply:
`cachedSymbols.get` returns `null` if the input argument is `null`

Reply:
Great, my bad then.